optimum/intel/neural_compressor/utils.py

```diff
@@ -109,3 +114,240 @@ def load_quantized_model(checkpoint_dir_or_file: str, model: torch.nn.Module, **
+    "XLNetLMHeadModelOutput": XLNetLMHeadModelOutput,
+}
+
+
+class AutoModelForGeneration(_BaseAutoModelClass):
+    _model_mapping = MODEL_FOR_CAUSAL_LM_MAPPING
+
+    # def forward(self, input_ids=None, past_key_values=None, attention_mask=None, **kwargs):
+    def __call__(self, *args, **kwargs):
+        past_key_values = getattr(kwargs, "past_key_values", None)
+        mems = getattr(kwargs, "mems", None)
+        if hasattr(self, "traced_model") and (past_key_values is not None or mems is not None):
+            inputs = []
+            kwargs.pop("position_ids", None)
+            for k, v in kwargs.items():
+                if v is not None and not isinstance(v, bool):
+                    inputs.append(v)
+            inputs = tuple(inputs)
+            out = self.traced_model(*inputs)
+            start_index = self.forward.__doc__.find("transformers.modeling_outputs.") + 30
+            end_index = self.forward.__doc__.find("`", start_index)
+            class_name = self.forward.__doc__[start_index:end_index]
+            if class_name == "CausalLMOutput":
+                if hasattr(kwargs, "labels") and kwargs["labels"] is not None:
+                    loss = out[0]
+                    lm_logits = out[1]
+                else:
+                    lm_logits = out[0]
+                    loss = None
+            elif class_name == "XLNetLMHeadModelOutput":
+                if hasattr(kwargs, "labels") and kwargs["labels"] is not None:
+                    loss = out[0]
+                    lm_logits = out[1]
+                    mems = out[2]
+                else:
+                    lm_logits = out[0]
+                    mems = out[1]
+                    loss = None
+            else:
+                if hasattr(kwargs, "labels") and kwargs["labels"] is not None:
+                    loss = out[0]
+                    lm_logits = out[1]
+                    past_key_values = out[2]
+                else:
+                    lm_logits = out[0]
+                    past_key_values = out[1]
+                    loss = None
+
+            if class_name == "CausalLMOutput":
+                return mapping[class_name](
+                    loss=loss,
+                    logits=lm_logits,
+                )
+            elif class_name == "XLNetLMHeadModelOutput":
+                return mapping[class_name](
+                    loss=loss,
+                    logits=lm_logits,
+                    mems=past_key_values,
+                )
+            else:
+                return mapping[class_name](
+                    loss=loss,
+                    logits=lm_logits,
+                    past_key_values=past_key_values,
+                )
+        else:
+            return self.forward(*args, **kwargs)
+
+    def set_traced_model(self, traced_model):
+        self.traced_model = traced_model
+
+    @classmethod
+    def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
```

Comment:
I think it would make sense to subclass [`OptimizedModel`](https://github.com/huggingface/optimum/blob/v1.7.3/optimum/modeling_base.py#L72), overwritting `_from_transformers` to trace the model